app/services/affiliate-slots.js

```diff
@@ -81,133 +87,132 @@ const checkMobileSystem = (unit) => {
-      .filter(u => !u.isBig);
-
-    return allUnits.length > 0 ? allUnits[0] : undefined;
-  },
-
-  /**
-   * Get all units that can be displayed on current wiki with given `title`
-   *
-   * @param {string} title
-   * @returns {AffiliateUnit[]}
-   */
-  getAllUnitsOnPage(title) {
-    return this.getAllUnits([
-      t => !t.disableOnPage,
-      t => checkFilter(t.page, title),
-    ]).filter(
-      u => !u.disableOnPage,
-    );
-  },
-
-  /**
-   * Get all units that can be displayed on current wiki
-   *
-   * @param {Function[]} filter functions for targeting
-   * @returns {AffiliateUnit[]}
+   * @returns {Targeting[]}
    */
-  getAllUnits(filters = []) {
-    let activeTargeting = targeting
-      // filter targeting by GEO cookie (country)
-      .filter(t => checkFilter(t.country, this.currentCountry))
+  _getTargetingOnSearch(query) {
+    return searchPageTargeting
       // filter targeting by vertical name
       .filter(t => checkFilter(t.vertical, this.currentVertical))
       // filter targeting by current wiki ID
-      .filter(t => checkFilter(t.wikiId, this.currentWikiId));
+      .filter(t => checkFilter(t.wikiId, this.currentWikiId))
+      // filter targeting for search `query`
+      .filter(t => checkFilter(t.query, query));
+  },
 
-    // apply additional criteria
-    filters.forEach((f) => {
-      activeTargeting = activeTargeting.filter(f);
+  fetchUnitForSearch(query, isBig = false) {
+    return new Promise((resolve, reject) => {
+      // check if we have possible units (we can fail early if we don't)
+      if (!this._getAvailableUnits()) {
+        resolve(undefined);
+      }
+
+      // get the units that fulfill the targeting on search
+      const targeting = this._getTargetingOnSearch(query);
+      const units = this._getUnitsWithTargeting(targeting)
+        // filter type of ad - isBig can be undefined, let's convert both to boolean 
+        .filter(u => !!u.isBig === !!isBig)
+        // filter units disabled on search page
+        .filter(u => !u.disableOnSearch);
+      
+      // fetch only the first unit if available
+      resolve(units.length > 0 ? units[0] : undefined);
     });
+  },
 
-    // extract unit names from available targeting
-    const availableUnitNames = activeTargeting
-      // grab only names of allowed units
-      .map(t => t.unit)
-      // flatten the array
-      .reduce((prev, curr) => prev.concat(curr), [])
-      // make the list unique
-      .filter(distinct);
-
-    // get all the units based on name and additional checks
-    const availableUnits = units
-      // filter only units that are actually allowed (because of targeting)
-      .filter(u => availableUnitNames.indexOf(u.name) > -1)
-      // check for mobile-system-specific units
-      .filter(u => checkMobileSystem(u))
-      // sort them according to the priority
-      .sort((a, b) => ((a.priority > b.priority) ? 1 : -1));
-
-    return availableUnits;
+  fetchUnitForPage(id, isBig = false) {
```

Comment:
What does `id` refer to? Can it be renamed to be more clear?

Reply:
e7fa558